src/Common/Core/Test/Match/MatchAny.cs

```diff
@@ -2,25 +2,58 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using FluentAssertions.Formatting;
 
 namespace Microsoft.Common.Core.Test.Match {
     public class MatchAny<T> : IEquatable<T> {
         public static readonly MatchAny<T> Instance = new MatchAny<T>();
 
-        public bool Equals(T other) {
-            return true;
+        private readonly Func<T, bool> _condition;
+        private readonly Func<string> _toString;
+        private readonly object _wrapped;
+
+        static MatchAny() {
+            FluentAssertions.Formatting.Formatter.AddFormatter(new Formatter());
+        }
+
+        private MatchAny(Func<T, bool> condition, object wrapped) {
+            _condition = condition;
+            _toString = wrapped.ToString;
+            _wrapped = wrapped;
+        }
+
+        public MatchAny() {
+            _condition = _ => true;
+            _toString = () => "<any>";
         }
 
-        public override bool Equals(object obj) {
-            return true;
+        public MatchAny(Expression<Func<T, bool>> condition) {
+            _condition = condition.Compile();
+            _toString = () => $"<satisfies {condition.ToString()}>";
         }
 
-        public override int GetHashCode() {
-            return 0;
+        public static MatchAny<T> ThatMatches<TOther>(IEquatable<TOther> match) {
+            return new MatchAny<T>(x => x is TOther && match.Equals((TOther)(object)x), match);
         }
 
-        public override string ToString() {
-            return "<any>";
+        public bool Equals(T other) =>
+            _condition(other);
+
+        public override bool Equals(object obj) =>
+            (obj == null || obj is T) && Equals((T)obj);
+
+        public override int GetHashCode() => 0;
+
+        public override string ToString() => _toString();
+
+        private class Formatter : IValueFormatter {
+            public bool CanHandle(object value) =>
+                (value as MatchAny<T>)?._wrapped != null;
+
+            public string ToString(object value, bool useLineBreaks, IList<object> processedObjects = null, int nestedPropertyLevel = 0) =>
+                FluentAssertions.Formatting.Formatter.ToString(((MatchAny<T>)value)._wrapped, useLineBreaks, processedObjects, nestedPropertyLevel);
```

Comment:
Because it needs to delegate to the formatter of the wrapped object, which is not necessarily the same as its `ToString`.